subsystem/src/calls/handle.rs

```diff
@@ -0,0 +1,166 @@
+        self.send_action(Action::Mut(Box::new(func)))
+    }
+
+    /// Submit an async procedure to be performed by the subsystem.
+    pub fn submit_async(
+        &self,
+        func: impl FnOnce(&T) -> BoxFuture<()> + Send + 'static,
+    ) -> Result<(), SubmissionError> {
+        self.send_action(Action::Ref(Box::new(func)))
+    }
+
+    /// Submit a procedure to be preformed by the susbsystem (mutable).
+    pub fn submit_mut(
+        &self,
+        func: impl FnOnce(&mut T) + Send + 'static,
+    ) -> Result<(), SubmissionError> {
+        self.send_action(Action::Mut(Box::new(move |subsys| {
+            Box::pin(async { func(subsys) })
+        })))
+    }
+
+    /// Submit a procedure to be preformed by the susbsystem.
+    pub fn submit(&self, func: impl FnOnce(&T) + Send + 'static) -> Result<(), SubmissionError> {
+        self.send_action(Action::Ref(Box::new(move |subsys| {
+            Box::pin(async { func(subsys) })
+        })))
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+pub struct Handle<T: ?Sized>(SubmitOnlyHandle<T>);
+
+impl<T: ?Sized> Clone for Handle<T> {
+    fn clone(&self) -> Self {
+        self.shallow_clone()
+    }
+}
+
+impl<T: ?Sized> ShallowClone for Handle<T> {
+    fn shallow_clone(&self) -> Self {
+        Self(self.0.shallow_clone())
+    }
+}
+
+impl<T: ?Sized + Send + Sync + 'static> Handle<T> {
+    /// Create a new subsystem handle.
+    pub(crate) fn new(submit_only_handle: SubmitOnlyHandle<T>) -> Self {
+        Self(submit_only_handle)
+    }
+
+    /// Get an equivalent [SubmitOnlyHandle].
+    pub fn as_submit_only(&self) -> &SubmitOnlyHandle<T> {
+        &self.0
+    }
+
+    /// Call an async procedure to the subsystem. Result has to be await-ed explicitly
+    pub fn call_async_mut<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&mut T) -> BoxFuture<R> + Send + 'static,
+    ) -> CallResult<R> {
+        let (rtx, rrx) = tokio::sync::oneshot::channel::<R>();
+        let result = self.0.submit_async_mut(move |subsys| {
+            Box::pin(async move {
+                if rtx.send(func(subsys).await).is_err() {
+                    log::trace!("Subsystem call (mut) result ignored");
+                }
+            })
+        });
+        CallResult::new(result.map(|()| CallResponse::new(rrx)))
+    }
+
+    /// Call an async procedure to the subsystem (immutable).
+    pub fn call_async<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&T) -> BoxFuture<R> + Send + 'static,
+    ) -> CallResult<R> {
+        let (rtx, rrx) = tokio::sync::oneshot::channel::<R>();
+        let result = self.0.submit_async(move |subsys| {
+            Box::pin(async move {
+                if rtx.send(func(subsys).await).is_err() {
+                    log::trace!("Subsystem call (mut) result ignored");
```

Comment:
Should the "(mut)" be here?

Reply:
It should not. Well spotted, thanks. Fixed.